### PR TITLE
Add batching in TokenClassificationPipeline

### PR DESCRIPTION
**Pull request by parakalan**
# What does this PR do?

Currently, the NER pipeline in transformers iterates through the list of input sentences and processes them, sequentially.
This PR adds batching support in the pipeline to decrease latency and use GPU more efficiently.

Relevant Issue :-  #11244 

### Without Batching (CPU)
Device: CPU
No. examples: 1000
Time taken: 283.27826976776123

Device: GPU
No. examples: 1000
Time taken: 17.89318561553955

Please check the benchmark gist [here](https://gist.github.com/parakalan/88b613ed4ca0001afb60448996f6b62a)

### Without Batching (CPU)
Device: CPU
No. examples: 1000
Batch Size: 512
Time taken: 121.81582999229431

Device: GPU
No. examples: 1000
Batch Size: 512
Time taken: 2.780881404876709

Please check the benchmark gist [here](https://gist.github.com/parakalan/f1fa25f25b8a70125145afbcbbeac85f)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. - https://github.com/huggingface/transformers/issues/11244
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by LysandreJik** [comment]
FYI there is also work done on this pipeline in https://github.com/huggingface/transformers/pull/10568 if you want to give it a look! It doesn't concern batching, however.

**Comment by parakalan** [comment]
Thanks, let me check that out.

**Comment by parakalan** [comment]
Please review this  @LysandreJik , @Narsil , @joshdevins

File: src/transformers/pipelines/token_classification.py [hunk 1]
```diff
@@ -30,6 +31,7 @@ class TokenClassificationArgumentHandler(ArgumentHandler):
 
     def __call__(self, inputs: Union[str, List[str]], **kwargs):
 
+        model_batch_size = kwargs.get("model_batch_size", 1)
```
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:34]: You are actually using this as a boolean, not an int.

File: src/transformers/pipelines/token_classification.py [hunk 2]
```diff
@@ -140,44 +142,45 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
...
-                    entities = self.model(tokens.data)[0][0].numpy()
-                    input_ids = tokens["input_ids"].numpy()[0]
-                else:
-                    with torch.no_grad():
-                        tokens = self.ensure_tensor_on_device(**tokens)
-                        entities = self.model(**tokens)[0][0].cpu().numpy()
-                        input_ids = tokens["input_ids"].cpu().numpy()[0]
+            if self.framework == "tf":
+                if model_batch_size > 1:
+                    warnings.warn("The `model_batch_size` argument is not supported for Tensorflow models. Ignoring")
```
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:167]: You're actually breaking tensorflow, right ? because you're not checking all sentences anymore.

**Review by Narsil — COMMENTED** [review]
I appreciate the intent of this PR, but I think it's most likely to hit users in their own foot without them even realizing. I'm disapproving this PR for that reason (and it actually raises questions about other pipelines implementations)

- Pipelines are aimed at inference, and in inference, batching is almost always detrimental. I've modified the proposed example with a slightly less nice example to display the problem:
https://gist.github.com/Narsil/ee5c09875e74fa6f018dc6d014f6c06c. 
```
Model batch size 1  
Device: CPU (i7-4790)
No. examples: 200
Time taken: 10.693870067596436
Model batch size 2
Device: CPU (batched)
No. examples: 200
Time taken: 10.754503965377808
Model batch size 2
Device: CPU (batched 2nd)
No. examples: 200
Time taken: 34.8472113609314
-------------------------------------------------- (GTX 970)
Model batch size 1
Device: GPU
No. examples: 200
Time taken: 0.7129480838775635
Model batch size 2
Device: GPU (batched)
No. examples: 200
Time taken: 0.708709716796875
Model batch size 2
Device: GPU (batched 2nd)
No. examples: 200
Time taken: 8.97895359992981

```
The core of the issue is the extra padding tokens created while running inference. Those are very bad for overall efficiency, and in a live system it is most likely to occur (from experience, I can say that it can be overwhelmingly bad). It's almost never worth it to attempt batching in a live production, unless you're very sure about the alignment problem.
- That being noted, other pipelines do sometimes do batching, and it **can** be effective if used properly, even if it is very hard at the pipeline level (because we're receiving strings and not TOKENS, which should be used for nice alignment)

**Comment by parakalan** [comment]
Closing this PR based on @Narsil's review. Thanks
